src/branch_and_prune.jl

```diff
@@ -5,11 +5,94 @@ export branch_and_prune, Bisection, Newton
 
 diam(x::Root) = diam(x.interval)
 
-
 Base.size(x::Interval) = (1,)
 
 isinterior{N}(X::IntervalBox{N}, Y::IntervalBox{N}) = all(isinterior.(X, Y))
 
+# contractors:
+"""
+    Contractor{F}
+
+    Abstract type for contractors.
+"""
+abstract type Contractor{F} end
+
+export Bisection, Newton
+
+struct Bisection{F} <: Contractor{F}
+    dimension::Int
+    f::F
+end
+
+Bisection{n}(::Type{Val{n}}, f) = Bisection(n, f)
+
+function (contractor::Bisection)(X)
+    image = contractor.f(X)
+
+    if !(contains_zero(image))
+        return :empty, X
+    end
+
+    return :unknown, X
+end
+
+
+struct Newton{F,FP,O} <: Contractor{F}
+    dimension::Int
+    f::F
+    fp::FP
+    op::O
+end
+
+function Newton(::Type{Val{1}}, f::Function)
+    f_prime = x -> ForwardDiff.derivative(f, x)
+    Newton(1, f, f_prime, N)
+end
+
+function Newton{n}(::Type{Val{n}}, f::Function)
+    f_prime = x -> ForwardDiff.jacobian(f, x)
+    Newton(n, f, f_prime, N)
+end
+
+function Newton(f_prime::Function)
+    NewtonConstructor(f_prime)
```

Comment:
I don't see the need for a separate `NewtonConstructor` object here?

Reply:
It allows to store the derivative which is needed to handle complex function the same way they are in the default case i.e. to turn the complex derivative into a real jacobian.

However it is true that `Newton(f_prime::Function)` could simply return `f_prime`, and we could dispatch on the fact it is a function rather than some custom type. But then if we implement another method possibly using the derivative (I believe Krawcyzk method allows it but I still have to check it in detail), we lose the information about what method it is.

Reply:
I definitively have to improve my "code explaining skills" :D

The short and hopefully clearer answer about the constructor is that it is the handiest solution I found to be able to
1 Track the expected contractor (in this case `Newton`).
2 Keep the provided derivative easily accessible for manipulation until the contractor is created.

The second part being only useful for complex functions.

Anyway, I just figured that this can be more elegantly done by moving the instancing of the contractor out of `branch_and_prune` to the `roots` function. So that `branch_and_prune` take a contractor instance rather than a contractor type (or constructor) as third parameter.

This will also resolve any concern about type stability I think.

And finally, no, I did not do any benchmark.